### PR TITLE
USB: DFU class rework

### PR DESCRIPTION
**Pull request by jfischer-no**
This patch moves USB DFU class driver to subsys/usb/class.

For the first the USB DFU class driver depends on DFU image
manager and partition layout and is limited to use as an
application for the bootloader. The driver fetches the
information about the flash, erase block size, write block
size and partitions offset from the DT now. The driver has
two interfaces associated with the two partitions "SLOT-0"
and "SLOT-1". The "SLOT-0" can only be read.

In the following work the class driver can be extended so
that it can be used from the bootloader and update a flash
region directly from the bootloader.

### CODE REVIEW

**Review by nashif — CHANGES_REQUESTED** [review]
what about removing

./samples/subsys/usb/dfu/src/usb_dfu.h ?

**Comment by jfischer-phytec-iot** [comment]
@nashif the header should be removed, i am on it, but I am looking for a interface to partition and boot manager first.

**Review by nashif — DISMISSED** [review]

**Comment by jfischer-phytec-iot** [comment]
@mbolivar can you please review

**Comment by jfischer-phytec-iot** [comment]
quark boards does not have flash partitions

File: include/usb/class/usb_dfu.h
Comment (finikorg) [inline include/usb/class/usb_dfu.h]: License text gets somehow screwed up with double '*' though I understand you have copied it from dfu sample, @nashif can we use general SPDX stuff here instead of full license?

File: samples/subsys/usb/dfu/src/main.c
Comment (finikorg) [inline samples/subsys/usb/dfu/src/main.c]: last empty line not needed

**Review by finikorg — COMMENTED** [review]
Looks OK, not tested.

**Review by mbolivar — COMMENTED** [review]
Does USB DFU really need three partitions (application A/B and a swap area for moving data around), just like MCUboot? I'm confused about how I can help review, as I'm not very familiar with the USB DFU protocol.

I.e. I thought USB DFU was generally used with a single bootloader that just overwrites a single application area, with no need for slot 1 or scratch. (The only USB DFU bootloader I've ever hacked on worked that way). What am I missing?

**Comment by jfischer-phytec-iot** [comment]
@mbolivar It depends on how one want to use it. Maybe someone would like to use it with the MCUboot. The swap partition is definitely not needed to be exported, I will remove it (the code has three alternative configurations).

**Comment by mbolivar** [comment]
> @mbolivar It depends on how one want to use it. Maybe someone would like to use it with the MCUboot.

So the idea would be something like:

1. mcuboot boots as stage 0
2. mcuboot runs second-stage DFU bootloader
3. use USB DFU to update application in some other partition, reset
4. mcuboot boots as stage 0
5. mcuboot runs second stage DFU bootloader, which uses a strap pin or some other mechanism to jump to new application

Or so? I'm just trying to work out the flow here. Are the split partitions mandatory?

> The swap partition is definitely not needed to be exported, I will remove it (the code has three alternative configurations).

**Comment by jfischer-phytec-iot** [comment]
@mbolivar I have revised it, it uses `sys/dfu` now. The download (write to the slot1) can be started by the application, so the USB-DFU interface is part of the application. The other option would be to use it in the bootloader, I have not looked deeper into mcuboot yet, but I expect to use USB-DFU in the bootloader needs more work.

**Comment by mbolivar** [comment]
> @mbolivar I have revised it, it uses sys/dfu now. The download (write to the slot1) can be started by the application, so the USB-DFU interface is part of the application. The other option would be to use it in the bootloader, I have not looked deeper into mcuboot yet, but I expect to use USB-DFU in the bootloader needs more work.

@jfischer-phytec-iot so I think you're saying that the boot partition has the USB DFU driver, and the slot1 partition has the application being updated, right?

File: samples/subsys/usb/dfu/dts.overlay
Comment (carlescufi) [inline samples/subsys/usb/dfu/dts.overlay]: You don't need this anymore, you only need to enable `CONFIG_BOOTLOADER_MCUBOOT` in the `.conf`

File: samples/subsys/usb/dfu/CMakeLists.txt
Comment (carlescufi) [inline samples/subsys/usb/dfu/CMakeLists.txt]: You don't need this anymore, you only need to enable `CONFIG_BOOTLOADER_MCUBOOT` in the `.conf`

**Review by carlescufi — DISMISSED** [review]

**Comment by carlescufi** [comment]
@mbolivar:
>so I think you're saying that the boot partition has the USB DFU driver, and the slot1 partition has the application being updated, right?

No, think what @jfischer-phytec-iot is saying is that this DFU works exactly like the mcumgr or LWM2M ones. The application in slot0 includes USB support and is able to receive an image and store it into slot 1, and then a reboot will swap the images.

Further down the line we might look into adding USB support into MCUboot, but right now this acts as a normal "push/background/in-app" type DFU like the ones we have.

File: samples/subsys/usb/dfu/dts.overlay
Comment (jfischer-phytec-iot) [reply samples/subsys/usb/dfu/dts.overlay]: I think I can remove it completely, it is only needed if it is part of MCUboot based bootloader.

Comment (carlescufi) [reply samples/subsys/usb/dfu/dts.overlay]: You need `CONFIG_BOOTLOADER_MCUBOOT` if you are building an application that is to be booted by MCUboot, so any app using this DFU mechanism.

Comment (jfischer-phytec-iot) [reply samples/subsys/usb/dfu/dts.overlay]: You are right

**Comment by carlescufi** [comment]
recheck

**Review by carlescufi — COMMENTED** [review]
This looks pretty good now. I will test with the nRF52840 and report back

**Comment by carlescufi** [comment]
@jfischer-phytec-iot Can you rebase after i merged @sundar-subramaniyan's last PR with 52840 fixes?

**Comment by jfischer-phytec-iot** [comment]
@carlescufi rebased

**Comment by carlescufi** [comment]
@jfischer-phytec-iot Tried testing on an nRF52840 but it doesn't enumerate the device. The `cdc_acm` sample works fine with that board. Is there something incompatible with that USB driver?

**Comment by carlescufi** [comment]
>@jfischer-phytec-iot Tried testing on an nRF52840 but it doesn't enumerate the device. The cdc_acm sample works fine with that board. Is there something incompatible with that USB driver?

Ignore that, I was flashing the wrong signed file, sorry for the noise

File: samples/subsys/usb/dfu/prj.conf
Comment (carlescufi) [inline samples/subsys/usb/dfu/prj.conf]: Remove this line it is not needed with `CONFIG_BOOTLOADER_MCUBOOT`

**Review by carlescufi — CHANGES_REQUESTED** [review]
For some reason the DFU process does not complete for me and I get no logs on the device side. On the host I get:

```
carles@zephyr-nordic ~/src/mcuboot/boot/zephyr/b (master)
$ sudo dfu-util --list
dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

Found Runtime: [2fe3:0100] ver=0011, devnum=19, cfg=1, intf=0, path="3-1.2", alt=0, name="UNKNOWN", serial="0.01"

carles@zephyr-nordic ~/src/mcuboot/boot/zephyr/b (master)
$ sudo dfu-util -D ~/zp.s.bin
dfu-util 0.9

Copyright 2005-2009 Weston Schmidt, Harald Welte and OpenMoko Inc.
Copyright 2010-2016 Tormod Volden and Stefan Schmidt
This program is Free Software and has ABSOLUTELY NO WARRANTY
Please report bugs to http://sourceforge.net/p/dfu-util/tickets/

dfu-util: Invalid DFU suffix signature
dfu-util: A valid DFU suffix will be required in a future dfu-util release!!!
Opening DFU capable USB device...
ID 2fe3:0100
Run-time device DFU version 0110
Claiming USB DFU Runtime Interface...
Determining device status: state = appIDLE, status = 0
Device really in Runtime Mode, send DFU detach request...
Resetting USB...
dfu-util: More than one DFU capable USB device found! Try `--list' and specify the serial number or disconnect all but one device
```

**Comment by jfischer-phytec-iot** [comment]
@carlescufi I've tested it mostly with FRDM-K64F, I'll take a look

**Comment by carlescufi** [comment]
@jfischer-phytec-iot tried it today again, and it doesn't enumerate properly actually. Sometimes I manage to enumerate but then the DFU fails.

I see things like:

```
***** BOOTING ZEPHYR OS v1.11.99 - BUILD: Mar 19 2018 15:39:36 *****
[MCUBOOT] [INF] main: Starting bootloader
[MCUBOOT] [INF] boot_status_source: Image 0: magic=unset, copy_done=0xff, image_ok=0xff
[MCUBOOT] [INF] boot_status_source: Scratch: magic=unset, copy_done=0x0, image_ok=0xff
[MCUBOOT] [INF] boot_status_source: Boot source: slot 0
[MCUBOOT] [INF] boot_swap_type: Swap type: none
[MCUBOOT] [INF] main: Bootloader chainload address offset: 0xc000
[MCUBOOT] [INF] main: Jumping to the first image slot
***** BOOTING ZEPHYR OS v1.11.99 - BUILD: Mar 19 2018 15:47:36 *****
[usb/dc] [ERR] access_fault_handler: USBD Access Fault!
[usb/dc] [ERR] access_fault_handler: USBD Access Fault!
[usb/dc] [ERR] access_fault_handler: USBD Access Fault!
[usb/dc] [ERR] access_fault_handler: USBD Access Fault!
```

**Comment by sundar-subramaniyan** [comment]
@carlescufi My 2 cents -

From the log you have pasted, towards the end, there's a trouble with USB enumeration:

> [usb/dc] [DBG] usbd_handle_state_change: USB detected
> [usb/dc] [DB[usb/dc] [DBG] usbd_handle_state_change: USB Powered
> [usb/dc] [ERR] access_fault_handler: USBD Access Fault!
> [general] [DBG] dfu_status_cb: USB device connected
> G] hf_clock_enable: HF clock start success
> [usb/dc] [ERR] access_fault_handler: USBD Access Fault!
> [usb/dc] [ERR] access_fault_handler: USBD Access Fault!
> [usb/dc] [ERR] access_fault_handler: USBD Access Fault!
> [general] [INF] main: This device supports USB DFU class.

I believe there's a reset happening somewhere during the DFU update that makes it run into this state.
The Access fault is happening because of the anomaly mentioned in the errata [No. 187. “USB cannot be enabled”](http://infocenter.nordicsemi.com/index.jsp?topic=%2Fcom.nordic.infocenter.nrf52840.EngB.errata%2Fdita%2Ferrata%2FnRF52840%2FEngineeringB%2Flatest%2Ferr_840.html)

**Comment by sundar-subramaniyan** [comment]
@carlescufi 

I was able to flash and boot the DFU sample merged with mcuboot.
When I issue dfu-util --download  I get this:

```
$ sudo dfu-util --alt 1 --download /tmp/hello.s.bin 
dfu-util 0.5

(C) 2005-2008 by Weston Schmidt, Harald Welte and OpenMoko Inc.
(C) 2010-2011 Tormod Volden (DfuSe support)
This program is Free Software and has ABSOLUTELY NO WARRANTY

dfu-util does currently only support DFU version 1.0

Opening DFU USB device... ID 2fe3:0100
Run-time device DFU version 0110
Claiming USB DFU Runtime Interface...
Determining device status: state = appIDLE, status = 0
Device really in Runtime Mode, send DFU detach request...
Resetting USB...
Opening USB Device...
Found DFU: [2fe3:0100] devnum=0, cfg=1, intf=0, alt=1, name="image-1"
Claiming USB DFU Interface...
Setting Alternate Setting #1 ...
Determining device status: state = dfuIDLE, status = 0
dfuIDLE, continuing
DFU mode device DFU version 0110
Device returned transfer size 64
No valid DFU suffix signature
Warning: File has no DFU suffix
bytes_per_hash=913
Copying data from PC to DFU device
Starting download: [dfu_download: libusb_control_transfer returned -7
Error during download
```

Is the "No valid DFU suffix signature" harmless?
I'll give it another try later.

**Comment by sundar-subramaniyan** [comment]
@carlescufi 

I tried again with USB_DEVICE log level set to 4 and I get this:
```
[general] [DBG] usb_handle_request: ** 1 **
[general] [DBG] dfu_class_handle_req: DFU_DNLOAD block 0, len 64, state 2
[general] [DBG] dfu_class_handle_req: DFU_DNLOAD start
```

After about few seconds, the download fails for some reason with following logs:

```
[general] [DBG] dfu_flash_write: bytes written 0x0
[general] [DBG] usb_handle_control_transfer: >> usb_data_to_host(2)
[usb/dc] [ERR] handle_ctrl_ep_data_state_events: invalid event 0 in data state for EP 0
```

It looks like Bulk data out handling has some issues in nRF52840.
I'll take another look later.

**Comment by sundar-subramaniyan** [comment]
@carlescufi 

I'm using the following command to boot the merged hex file to boot the DFU sample (in case if anyone is wondering):

`nrfjprog --eraseall -f nrf52 && nrfjprog --program zephyr.hex.merged -f nrf52 && nrfjprog -r -f nrf52 && sleep 1 && nrfjprog -f nrf52 -p`

The sleep and pin reset is to workaround the USB power up anomaly from nRF52840 errata B.

I added some logs in usb_dfu.c and in mcuboot.c in boot_erase_img_bank() and I could see the flash_erase() is stuck for about ~9 seconds. Here's the log:

```
[general] [DBG] dfu_class_handle_req: DFU_DNLOAD block 0, len 64, state 2
[general] [DBG] dfu_class_handle_req: DFU_DNLOAD start
[general] [DBG] dfu_class_handle_req: reset counters done
[general] [DBG] dfu_class_handle_req: boot erase image start
write prot off
flash erase size: 430080 uptime: 8150
flash erase complete uptime: 17000
write prot on
[general] [DBG] dfu_class_handle_req: boot erase image end
[general] [DBG] dfu_class_handle_req: DNLOAD IDLE. dnload size: 64
[general] [DBG] dfu_flash_write: bytes written 0x0
[general] [DBG] usb_handle_control_transfer: >> usb_data_to_host(2)
[usb/dc] [ERR] handle_ctrl_ep_data_state_events: invalid event 0 in data state for EP 0
```

The first argument after "flash erase" debug print is the erase size and second argument is uptime in milliseconds. The next debug print "flash erase complete" is with uptime in milliseconds. The difference is 8850 ms. For an erase size of 420K, is this normal?

I'm not an expert in DFU but I believe the dfu-util times out waiting for a response.
If this is synchronous call, I guess the wait time should be expected time of "flash erase + flash write of first 64 bytes".

**Comment by carlescufi** [comment]
@sundar-subramaniyan 

>I believe there's a reset happening somewhere during the DFU update that makes it run into this state.

No, my enumeration problems happen just after flashing the DFU sample, well before I have a chance to run `dfu-util`. The sample itself doesn't enumerate, right after flashing with nrfjprog

**Comment by carlescufi** [comment]
@sundar-subramaniyan I think we are seeing different problems. In my case the initial enumeration after flashing with nrfjprog works fine with the `cdc_acm` sample but fails with this one. That is the stage that I'm trying to get past before I can really try DFU

**Comment by carlescufi** [comment]
@nvlsianpu Can you take a look at [this](https://github.com/zephyrproject-rtos/zephyr/pull/4373#issuecomment-374469640) comment and reply regarding the flash erase time?

**Comment by nvlsianpu** [comment]
@carlescufi @sundar-subramaniyan 
One page erase can take like 89 ms:
(https://github.com/zephyrproject-rtos/zephyr/blob/master/arch/arm/soc/nordic_nrf/nrf52/soc.h#L240)

105 page of flash are erased - so it can take up to 9,4185 s. So time measured is in this limit.
Time of writing of 64 bytes is neglible in compassion with time of erase such flash area.

**Comment by sundar-subramaniyan** [comment]
@nvlsianpu Thanks for the confirmation on the erase times.

**Comment by sundar-subramaniyan** [comment]
@jfischer-phytec-iot I had a first look at the captures you have uploaded. In the case of nRF52, there's supposed to be an OUT transaction (Status stage) followed by IN transaction and is missing. I'll take another look later and debug it. Thanks for the captures.

File: samples/subsys/usb/dfu/prj.conf
Comment (carlescufi) [reply samples/subsys/usb/dfu/prj.conf]: @jfischer-phytec-iot Can you remove this line so we can get this merged? Looks like there is no issue with DFU itself, but rather with the Nordic driver.

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: samples/subsys/usb/dfu/prj.conf
Comment (jfischer-phytec-iot) [reply samples/subsys/usb/dfu/prj.conf]: I will do, but I would like to try the page by page erase with NRF52 SoC before this PR will be merged.

**Comment by sundar-subramaniyan** [comment]
@jfischer-phytec-iot 

As you pointed out, nRF52840 doesn't ACK the OUT transaction (status stage) when the setup->wLength is exactly 64 bytes, because I thought the usb device stack would take care of writing a ZLP to EP0 IN.
Please see [this line](https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/usb/device/usb_dc_nrf5.c#L1239).

There are these following cases in my understanding for control read with data stage with read length (wLength):
    a) Less than 64 bytes (single IN transaction)
    b) Exactly 64 bytes (single IN transaction)
    c) More than 64 bytes (multiple IN transactions)

Now my question is, whose job is it to initiate status stage - USB stack or USB driver?

For case a) The USB driver can implicitly know from short packet length and initiate status stage. This is taken care already by the nRF5 USB driver.
For cases b) and c) how does the driver implicitly understand when to initiate status stage without storing the wLength in the driver context and decrementing every IN transaction? It has to be taken care by the USB stack, no?

Just to debug, I made [this change](https://gist.github.com/sundar-subramaniyan/e85bd07549473a69c99bae1de64864f5) to initiate status stage even if the length wrote is 64 bytes in write completion in the USB driver and I get the image uploaded to host PC. All 420KB of image_1.

Although, at the end of upload I get this error:
`dfu_upload: libusb_control_msg returned -9`

Looking at libusb source, -9 corresponds to LIBUSB_ERROR_PIPE.
Looking at the wireshark, I could see the "SET INTERFACE REQUEST" and "SET INTERFACE RESPONSE" after all the "CONTROL IN" transactions (6720 blocks).
The "SET INTERFACE" sets alternate interface 0 which means it's switching back the interface after upload is complete.

Is the error print normal?

**Comment by mbolivar** [comment]
> @mbolivar: @carlescufi explained it correctly . I just do not know how to configure it for Quark SE based boards, because they do not have flash partitions.

OK, thanks for the confirmation!

Since these changes are all in the USB subsystem, I think there's nothing for me to review here and will leave you all to it.

**Comment by sundar-subramaniyan** [comment]
@jfischer-phytec-iot 

> About ZLP, refer to usb spec. 8.5.3.2 "Variable-length Data Stage", what is not the case if the setup packet requests for 64 bytes. Also "Variable-length Data Stage" should be handled in the device stack.

I referred to the USB spec section 8.5.3.2 and here is the excerpt:

`A control pipe may have a variable-length data phase in which the host requests more data than is contained in the specified data structure. When all of the data structure is returned to the host, the function should indicate that the Data stage is ended by returning a packet that is shorter than the MaxPacketSize for the pipe. If the data structure is an exact multiple of wMaxPacketSize for the pipe, the function will return a zero-length packet to indicate the end of the Data stage.`

This is exactly what the nRF USB driver from master does - For short packets or ZLP packets, driver initiates (prepares) HW for status stage. Until this happens, the HW shall NACK the OUT transactions (status stage) in the case of control read. In DFU upload, after 64 bytes of transfer, the USB stack does not write a ZLP and so driver never prepares the HW for the status stage and so it gets stuck during the first control read transaction. What I expect is that, after the write completion of the 64 bytes, the stack should check if the last transaction packet size is 64 bytes and should write ZLP to end the transaction.

You can see [this line](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/usb_device.c#L326) where the USB stack as of now checks if all the  buffer contents are written and if so it does not write ZLP.

Am I missing something here?

**Comment by sundar-subramaniyan** [comment]
@jfischer-phytec-iot 

I made [this change](https://gist.github.com/sundar-subramaniyan/7da50a34c31473e837de809340784c1b) to the USB stack and removed my [earlier change](https://gist.github.com/sundar-subramaniyan/e85bd07549473a69c99bae1de64864f5) in nRF USB driver and things seem to work similar way for DFU upload. Please see if this makes sense.

**Comment by sundar-subramaniyan** [comment]
@jfischer-phytec-iot For DFU download, I made some changes to nRF USB driver and USB DFU class driver as well to get it barely working. This is besides the change I did for upload in the USB device stack.

[nRF USB driver change](https://gist.github.com/sundar-subramaniyan/890d0e97b76a5f6b04ed8a8168055610)
[USB DFU class change](https://gist.github.com/sundar-subramaniyan/d076f41b601b9dee2a45e99ceee672c2)

These are work in progress changes and I need to tweak the work delay to get the download faster.
Right now it's hard coded to 1 second (for each chunk!!). What do you think about these changes?

I'll make a PR once I wrap up all these changes together.

**Comment by carlescufi** [comment]
Our USB driver needs a bit of rework, but I think this is independent. @jfischer-phytec-iot could you rebase please to see if we can merge this ahead of the driver work?

**Comment by jfischer-phytec-iot** [comment]
@carlescufi rebased

File: samples/subsys/usb/dfu/README.rst
Comment (dbkinder) [inline samples/subsys/usb/dfu/README.rst]: change to:

    Your SOC must run MCUboot as the stage 1 bootloader.  This sample
    is built as an application for the MCUboot bootloader.

File: samples/subsys/usb/dfu/README.rst
Comment (dbkinder) [inline samples/subsys/usb/dfu/README.rst]: Change to:

    Use the ``scripts/imagetool.py`` script from the `MCUboot GitHub repo`_  
    to sign the image.  (See the `Using MCUboot with Zephyr`_ documentation for
    details.)

.. _MCUboot GitHub repo: https://github.com/runtimeco/mcuboot
.. _Using MCUboot with Zephyr: https://mcuboot.com/mcuboot/readme-zephyr.html

File: samples/subsys/usb/dfu/README.rst
Comment (dbkinder) [inline samples/subsys/usb/dfu/README.rst]: change to:

    the `Using MCUboot with Zephyr`_ documentation.  Then build, sign, and flash

File: samples/subsys/usb/dfu/README.rst
Comment (dbkinder) [inline samples/subsys/usb/dfu/README.rst]: Change to:

    The Linux ``dfu-util`` tool ....

File: samples/subsys/usb/dfu/README.rst
Comment (dbkinder) [inline samples/subsys/usb/dfu/README.rst]: Make two sentences:

    Reset the SoC.  MCU boot will swap the images and boot
    the new application, showing this output to the console:

File: samples/subsys/usb/dfu/README.rst
Comment (dbkinder) [inline samples/subsys/usb/dfu/README.rst]: What console output will I see on this next reboot?

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: samples/subsys/usb/dfu/README.rst
Comment (jfischer-phytec-iot) [reply samples/subsys/usb/dfu/README.rst]: Yes, my mistake, I have fixed (and tested) it now.

File: samples/subsys/usb/dfu/README.rst
Comment (dbkinder) [inline samples/subsys/usb/dfu/README.rst]: ``as the stage``

File: samples/subsys/usb/dfu/README.rst
Comment (dbkinder) [inline samples/subsys/usb/dfu/README.rst]: ``built``

File: samples/subsys/usb/dfu/README.rst
Comment (dbkinder) [inline samples/subsys/usb/dfu/README.rst]: Don't need to say "please".  Add a blank line to start a new paragraph and say,

    Enable the MCUboot Kconfig option (as described in :ref:`mcuboot`) by
    adding the following line to your local copy of :file:`samples/hello_world/prj.conf`:

Actually, isn't there a way to do this with a project conf file rather than making
changes to the sample's prj.conf file?

**Review by dbkinder — CHANGES_REQUESTED** [review]
Looking better, thanks!

File: samples/subsys/usb/dfu/README.rst
Comment (jfischer-phytec-iot) [reply samples/subsys/usb/dfu/README.rst]: prj.conf should be changed before a second application is built, I think there should not be a blank line in between.

Comment (carlescufi) [reply samples/subsys/usb/dfu/README.rst]: @dbkinder "Please" has been removed and I think @jfischer-phytec-iot 's point is valid.

**Review by carlescufi — APPROVED** [review]

File: samples/subsys/usb/dfu/README.rst
Comment (dbkinder) [inline samples/subsys/usb/dfu/README.rst]: ``Do not``

**Review by dbkinder — CHANGES_REQUESTED** [review]
Ok with me.  One last typo...

File: samples/subsys/usb/dfu/README.rst
Comment (jfischer-phytec-iot) [reply samples/subsys/usb/dfu/README.rst]: @dbkinder thanks for review

**Review by dbkinder — APPROVED** [review]

**Comment by carlescufi** [comment]
@finikorg any objections to merging this?

**Pull request merged**
